fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/plans/InSource.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * InSource.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2021 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.query.plan.plans;
+
+import com.apple.foundationdb.record.EvaluationContext;
+import com.apple.foundationdb.record.ObjectPlanHash;
+import com.apple.foundationdb.record.PlanHashable;
+import com.apple.foundationdb.record.query.plan.temp.Quantifier;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+
+public abstract class InSource implements PlanHashable {
+    @SuppressWarnings("unchecked")
+    private static final Comparator<Object> VALUE_COMPARATOR = (o1, o2) -> ((Comparable)o1).compareTo((Comparable)o2);
+
+    @Nonnull
+    private final String bindingName;
+
+    protected InSource(@Nonnull final String bindingName) {
+        this.bindingName = bindingName;
+    }
+
+    @Nonnull
+    public String getBindingName() {
+        return bindingName;
+    }
+
+    public abstract boolean isSorted();
+
+    public abstract boolean isReverse();
+
+    protected abstract int size(@Nonnull EvaluationContext context);
+
+    @Nonnull
+    public List<Object> getValues() {
+        return getValues(null);
+    }
+
+    @Nonnull
+    protected abstract List<Object> getValues(@Nullable EvaluationContext context);
+
+    @Nonnull
+    public abstract RecordQueryInJoinPlan toInJoinPlan(@Nonnull final Quantifier.Physical innerQuantifier);
+
+    public int baseHash(@Nonnull final PlanHashKind hashKind, @Nonnull ObjectPlanHash objectPlanHash) {
```

Comment:
No -- it's not. This code taken from `RecordQueryInUnionPlan`.

https://github.com/FoundationDB/fdb-record-layer/blob/da38c80d5addfec943f0d3957c5cb3630e957ec0/fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/plans/RecordQueryInUnionPlan.java#L298

This is really unfortunate but I don't think we can just add that here as this will regress all kinds of plan hashes.

I will add a big todo here. We need to roll the plan hashes in a coordinated manner anyway. This should be done then as well.

Reply:
also, we can *not* use this method for the cascades planner (have it issue a different hash)